### PR TITLE
Pass correct HirId to late_bound_vars in diagnostic code

### LINKED ISSUE #125655: `No bound vars found for HirId(DefId(` in compiler/rustc_middle/src/ty/context.rs
### Code

```Rust
fn main() {
    static foo: dyn Fn() -> u32 = || -> u32 {
        //~^ ERROR the size for values of type
        //~| ERROR cannot be shared between threads safely
        //~| ERROR the size for values of type
        ...
        0
    };
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.80.0-nightly (84b40fc90 2024-05-27)
binary: rustc
commit-hash: 84b40fc908c3adc7e0e470b3fbaa264df0e122b8
commit-date: 2024-05-27
host: x86_64-apple-darwin
release: 1.80.0-nightly
LLVM version: 18.1.6
```

### Error output

```
info: syncing channel updates for 'nightly-x86_64-apple-darwin'

  nightly-x86_64-apple-darwin unchanged - rustc 1.80.0-nightly (84b40fc90 2024-05-27)

info: checking for self-update
error: unexpected token: `...`
 --> ./6FA6E1FD2230DECFC4A324415F1CE359CF44F0600A6F19AA26DE8E5404A6DAB9.rs:6:9
  |
6 |         ...
  |         ^^^
  |
help: use `..` for an exclusive range
  |
6 |         ..
  |
help: or `..=` for an inclusive range
  |
6 |         ..=
  |

error[E0277]: `(dyn Fn() -> u32 + 'static)` cannot be shared between threads safely
 --> ./6FA6E1FD2230DECFC4A324415F1CE359CF44F0600A6F19AA26DE8E5404A6DAB9.rs:2:17
  |
2 |     static foo: dyn Fn() -> u32 = || -> u32 {
  |                 ^^^^^^^^^^^^^^^ `(dyn Fn() -> u32 + 'static)` cannot be shared between threads safely
  |
  = help: the trait `Sync` is not implemented for `(dyn Fn() -> u32 + 'static)`
  = note: shared static variables must have a type that implements `Sync`

error[E0277]: the size for values of type `(dyn Fn() -> u32 + 'static)` cannot be known at compilation time
 --> ./6FA6E1FD2230DECFC4A324415F1CE359CF44F0600A6F19AA26DE8E5404A6DAB9.rs:2:17
  |
2 |     static foo: dyn Fn() -> u32 = || -> u32 {
  |                 ^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `(dyn Fn() -> u32 + 'static)`

error[E0277]: the size for values of type `(dyn Fn() -> u32 + 'static)` cannot be known at compilation time
 --> ./6FA6E1FD2230DECFC4A324415F1CE359CF44F0600A6F19AA26DE8E5404A6DAB9.rs:2:35
  |
2 |       static foo: dyn Fn() -> u32 = || -> u32 {
  |  ___________________________________^
3 | |         //~^ ERROR the size for values of type
4 | |         //~| ERROR cannot be shared between threads safely
5 | |         //~| ERROR the size for values of type
6 | |         ...
7 | |         0
8 | |     };
  | |_____^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `(dyn Fn() -> u32 + 'static)`
  = note: constant expressions must have a statically known size
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: compiler/rustc_middle/src/ty/context.rs:2583:21: No bound vars found for HirId(DefId(0:4 ~ 6FA6E1FD2230DECFC4A324415F1CE359CF44F0600A6F19AA26DE8E5404A6DAB9[8430]::main::foo).0) (static foo)

thread 'rustc' panicked at compiler/rustc_middle/src/ty/context.rs:2583:21:
Box<dyn Any>
stack backtrace:
   0:        0x10ac9c193 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h51921f2e52d686b8
   1:        0x10ace77cb - core::fmt::write::h2b73508d07b1ddda
   2:        0x10ac9224e - std::io::Write::write_fmt::h1a4990725cb94d19
   3:        0x10ac9bf81 - std::sys_common::backtrace::print::hbaa9d5d0b6ae5771
   4:        0x10ac9ed49 - std::panicking::default_hook::{{closure}}::hb29728a481024744
   5:        0x10ac9eac8 - std::panicking::default_hook::h863a389bd8aa7cb6
   6:        0x113a2db1c - std[1da7d9055ec82a98]::panicking::update_hook::<alloc[d2ff4adad34ee5b0]::boxed::Box<rustc_driver_impl[d3ad18f46d0afcfd]::install_ice_hook::{closure#0}>>::{closure#0}
   7:        0x10ac9f853 - std::panicking::rust_panic_with_hook::h2d23fde0880a2e81
   8:        0x113a981ba - std[1da7d9055ec82a98]::panicking::begin_panic::<rustc_errors[6b485a6418fcb1e4]::ExplicitBug>::{closure#0}
   9:        0x113a8fe59 - std[1da7d9055ec82a98]::sys_common::backtrace::__rust_end_short_backtrace::<std[1da7d9055ec82a98]::panicking::begin_panic<rustc_errors[6b485a6418fcb1e4]::ExplicitBug>::{closure#0}, !>
  10:        0x118228039 - std[1da7d9055ec82a98]::panicking::begin_panic::<rustc_errors[6b485a6418fcb1e4]::ExplicitBug>
  11:        0x113aab326 - <rustc_errors[6b485a6418fcb1e4]::diagnostic::BugAbort as rustc_errors[6b485a6418fcb1e4]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  12:        0x1145c392e - rustc_middle[4fd479a6026b06e9]::util::bug::opt_span_bug_fmt::<rustc_span[5c871bc8a1860ee8]::span_encoding::Span>::{closure#0}
  13:        0x114574a07 - rustc_middle[4fd479a6026b06e9]::ty::context::tls::with_opt::<rustc_middle[4fd479a6026b06e9]::util::bug::opt_span_bug_fmt<rustc_span[5c871bc8a1860ee8]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  14:        0x1145744a5 - rustc_middle[4fd479a6026b06e9]::ty::context::tls::with_context_opt::<rustc_middle[4fd479a6026b06e9]::ty::context::tls::with_opt<rustc_middle[4fd479a6026b06e9]::util::bug::opt_span_bug_fmt<rustc_span[5c871bc8a1860ee8]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  15:        0x1182f289b - rustc_middle[4fd479a6026b06e9]::util::bug::bug_fmt
  16:        0x11458d73d - <rustc_middle[4fd479a6026b06e9]::ty::context::TyCtxt>::late_bound_vars
  17:        0x113fc5845 - <rustc_hir_typeck[70f6020ddbcdf028]::fn_ctxt::FnCtxt>::suggest_missing_return_type
  18:        0x113ebec2d - <rustc_hir_typeck[70f6020ddbcdf028]::coercion::CoerceMany<rustc_hir[187abc7f04de6ccd]::hir::Expr>>::report_return_mismatched_types
  19:        0x113f136ba - <rustc_hir_typeck[70f6020ddbcdf028]::fn_ctxt::FnCtxt>::check_block_with_expected
  20:        0x113ed3e7b - <rustc_hir_typeck[70f6020ddbcdf028]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  21:        0x113ed5ec4 - <rustc_hir_typeck[70f6020ddbcdf028]::fn_ctxt::FnCtxt>::check_return_expr
  22:        0x113f94950 - rustc_hir_typeck[70f6020ddbcdf028]::check::check_fn
  23:        0x113fb6f62 - <rustc_hir_typeck[70f6020ddbcdf028]::fn_ctxt::FnCtxt>::check_expr_kind
  24:        0x113ed3e7b - <rustc_hir_typeck[70f6020ddbcdf028]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  25:        0x113f8c86b - rustc_hir_typeck[70f6020ddbcdf028]::typeck
  26:        0x114ef755a - rustc_query_impl[3691b0e6aab602df]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3691b0e6aab602df]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[4fd479a6026b06e9]::query::erase::Erased<[u8; 8usize]>>
  27:        0x114d6e607 - rustc_query_system[cc1e6c2a417942e6]::query::plumbing::try_execute_query::<rustc_query_impl[3691b0e6aab602df]::DynamicConfig<rustc_query_system[cc1e6c2a417942e6]::query::caches::VecCache<rustc_hir[187abc7f04de6ccd]::hir_id::OwnerId, rustc_middle[4fd479a6026b06e9]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[3691b0e6aab602df]::plumbing::QueryCtxt, false>
  28:        0x114f19abb - rustc_query_impl[3691b0e6aab602df]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  29:        0x113ccb2de - rustc_hir_analysis[5e381ea6a7b23192]::check::check::check_item_type
  30:        0x113cff284 - rustc_hir_analysis[5e381ea6a7b23192]::check::wfcheck::check_well_formed
  31:        0x114ef484a - rustc_query_impl[3691b0e6aab602df]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3691b0e6aab602df]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[4fd479a6026b06e9]::query::erase::Erased<[u8; 1usize]>>
  32:        0x114d6c26c - rustc_query_system[cc1e6c2a417942e6]::query::plumbing::try_execute_query::<rustc_query_impl[3691b0e6aab602df]::DynamicConfig<rustc_query_system[cc1e6c2a417942e6]::query::caches::VecCache<rustc_hir[187abc7f04de6ccd]::hir_id::OwnerId, rustc_middle[4fd479a6026b06e9]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[3691b0e6aab602df]::plumbing::QueryCtxt, false>
  33:        0x114f3ae10 - rustc_query_impl[3691b0e6aab602df]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  34:        0x113c41c92 - rustc_middle[4fd479a6026b06e9]::query::plumbing::query_ensure_error_guaranteed::<rustc_query_system[cc1e6c2a417942e6]::query::caches::VecCache<rustc_hir[187abc7f04de6ccd]::hir_id::OwnerId, rustc_middle[4fd479a6026b06e9]::query::erase::Erased<[u8; 1usize]>>, ()>
  35:        0x113d0da21 - rustc_hir_analysis[5e381ea6a7b23192]::check::wfcheck::check_mod_type_wf
  36:        0x114ef483a - rustc_query_impl[3691b0e6aab602df]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3691b0e6aab602df]::query_impl::check_mod_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[4fd479a6026b06e9]::query::erase::Erased<[u8; 1usize]>>
  37:        0x114d225a1 - rustc_query_system[cc1e6c2a417942e6]::query::plumbing::try_execute_query::<rustc_query_impl[3691b0e6aab602df]::DynamicConfig<rustc_query_system[cc1e6c2a417942e6]::query::caches::DefaultCache<rustc_span[5c871bc8a1860ee8]::def_id::LocalModDefId, rustc_middle[4fd479a6026b06e9]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[3691b0e6aab602df]::plumbing::QueryCtxt, false>
  38:        0x114f18f10 - rustc_query_impl[3691b0e6aab602df]::query_impl::check_mod_type_wf::get_query_non_incr::__rust_end_short_backtrace
  39:        0x113d8a57c - rustc_hir_analysis[5e381ea6a7b23192]::check_crate
  40:        0x114210011 - rustc_interface[3f7acd64cf69f8d0]::passes::run_required_analyses
  41:        0x114212653 - rustc_interface[3f7acd64cf69f8d0]::passes::analysis
  42:        0x114ef75aa - rustc_query_impl[3691b0e6aab602df]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[3691b0e6aab602df]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[4fd479a6026b06e9]::query::erase::Erased<[u8; 1usize]>>
  43:        0x114cd894e - rustc_query_system[cc1e6c2a417942e6]::query::plumbing::try_execute_query::<rustc_query_impl[3691b0e6aab602df]::DynamicConfig<rustc_query_system[cc1e6c2a417942e6]::query::caches::SingleCache<rustc_middle[4fd479a6026b06e9]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[3691b0e6aab602df]::plumbing::QueryCtxt, false>
  44:        0x114f01767 - rustc_query_impl[3691b0e6aab602df]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  45:        0x1139de277 - <rustc_interface[3f7acd64cf69f8d0]::queries::QueryResult<&rustc_middle[4fd479a6026b06e9]::ty::context::GlobalCtxt>>::enter::<core[917b7ec437872f2c]::result::Result<(), rustc_span[5c871bc8a1860ee8]::ErrorGuaranteed>, rustc_driver_impl[d3ad18f46d0afcfd]::run_compiler::{closure#0}::{closure#1}::{closure#3}>
  46:        0x113a34ced - rustc_interface[3f7acd64cf69f8d0]::interface::run_compiler::<core[917b7ec437872f2c]::result::Result<(), rustc_span[5c871bc8a1860ee8]::ErrorGuaranteed>, rustc_driver_impl[d3ad18f46d0afcfd]::run_compiler::{closure#0}>::{closure#1}
  47:        0x113a1f1f1 - std[1da7d9055ec82a98]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[3f7acd64cf69f8d0]::util::run_in_thread_with_globals<rustc_interface[3f7acd64cf69f8d0]::util::run_in_thread_pool_with_globals<rustc_interface[3f7acd64cf69f8d0]::interface::run_compiler<core[917b7ec437872f2c]::result::Result<(), rustc_span[5c871bc8a1860ee8]::ErrorGuaranteed>, rustc_driver_impl[d3ad18f46d0afcfd]::run_compiler::{closure#0}>::{closure#1}, core[917b7ec437872f2c]::result::Result<(), rustc_span[5c871bc8a1860ee8]::ErrorGuaranteed>>::{closure#0}, core[917b7ec437872f2c]::result::Result<(), rustc_span[5c871bc8a1860ee8]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[917b7ec437872f2c]::result::Result<(), rustc_span[5c871bc8a1860ee8]::ErrorGuaranteed>>
  48:        0x113a3ba96 - <<std[1da7d9055ec82a98]::thread::Builder>::spawn_unchecked_<rustc_interface[3f7acd64cf69f8d0]::util::run_in_thread_with_globals<rustc_interface[3f7acd64cf69f8d0]::util::run_in_thread_pool_with_globals<rustc_interface[3f7acd64cf69f8d0]::interface::run_compiler<core[917b7ec437872f2c]::result::Result<(), rustc_span[5c871bc8a1860ee8]::ErrorGuaranteed>, rustc_driver_impl[d3ad18f46d0afcfd]::run_compiler::{closure#0}>::{closure#1}, core[917b7ec437872f2c]::result::Result<(), rustc_span[5c871bc8a1860ee8]::ErrorGuaranteed>>::{closure#0}, core[917b7ec437872f2c]::result::Result<(), rustc_span[5c871bc8a1860ee8]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[917b7ec437872f2c]::result::Result<(), rustc_span[5c871bc8a1860ee8]::ErrorGuaranteed>>::{closure#2} as core[917b7ec437872f2c]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  49:        0x10aca8f0b - std::sys::pal::unix::thread::Thread::new::thread_start::hff1aeae30732a5f5
  50:     0x7ff801f5318b - __pthread_start

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/Documents/rustc-ice-2024-05-28T12_02_57-74152.txt` to your bug report

query stack during panic:
#0 [typeck] type-checking `main::foo`
#1 [check_well_formed] checking that `main::foo` is well-formed
end of query stack
error: aborting due to 5 previous errors
```

</p>
</details>

### LINKED ISSUE #130391: ICE: `Not enough bound vars`
auto-reduced (treereduce-rust):
````rust
fn main() {
    let _ = for<'a> |x: &u8| -> &'a () { x };
}


````

<details><summary><strong>original code</strong></summary>
<p>

original:
````rust
#![feature(closure_lifetime_binder)]

fn main() {
    // Implicit types
    let _ = for<> || {};                                      //~ ERROR implicit types in closure signatures are forbidden when `for<...>` is present
    let _ = for<'a> || -> &'a _ { &() };                      //~ ERROR implicit types in closure signatures are forbidden when `for<...>` is present
    let _ = for<'a> |x| -> &'a () { x };                      //~ ERROR implicit types in closure signatures are forbidden when `for<...>` is present
    let _ = for<'a> |x: &'a _| -> &'a () { x };               //~ ERROR implicit types in closure signatures are forbidden when `for<...>` is present
    let _ = for<'a> |x: &'a Vec::<_>| -> &'a Vec::<()> { x }; //~ ERROR implicit types in closure signatures are forbidden when `for<...>` is present
    let _ = for<'a> |x: &'a Vec<()>| -> &'a Vec<_> { x };     //~ ERROR implicit types in closure signatures are forbidden when `for<...>` is present
    let _ = for<'a> |x: &'a _| -> &'a &'a () { x };           //~ ERROR implicit types in closure signatures are forbidden when `for<...>` is present
    let _ = for<'a> |x: &'a _, y, z: _| -> &'a _ {            //~ ERROR implicit types in closure signatures are forbidden when `for<...>` is present
        let _: &u8 = x;
        let _: u32 = y;
        let _: i32 = z;
        x
    };

    // Lifetime elision
    let _ = for<> |_: &()| -> () {};           //~ ERROR `&` without an explicit lifetime name cannot be used here
    let _ = for<> |x: &()| -> &() { x };       //~ ERROR `&` without an explicit lifetime name cannot be used here
                                               //~| ERROR `&` without an explicit lifetime name cannot be used here
    let _ = for<> |x: &'_ ()| -> &'_ () { x }; //~ ERROR `'_` cannot be used here
                                               //~| ERROR `'_` cannot be used here
    let _ = for<'a> |x: &u8| -> &'a () { x };  //~ ERROR `&` without an explicit lifetime name cannot be used here
    let _ = for<'a> |x: &'a ()| -> &() { x };  //~ ERROR `&` without an explicit lifetime name cannot be used here
    //~^ ERROR: lifetime may not live long enough
}

````
</p>
</details>

Version information
````
rustc 1.83.0-dev
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.83.0-dev
LLVM version: 19.1.0

````

@rustbot label +F-closure_lifetime_binder

### LINKED ISSUE #130663: ICE: closure_lifetime_binder: BrNamed != BrNamed
auto-reduced (treereduce-rust):
````rust
fn hello(x: &Ty) {
    let _ = for<'a> |x: &u8| -> &'a () { x };
}

struct Ty;

fn A() {}


````

original:
````rust
#![feature(async_closure)]

fn hello(x: &Ty) {
    let _ = for<'a> |x: &u8| -> &'a () { x };
}

struct Ty;
impl Ty {
    fn hello(&self) {}
}

fn A() {}

````

Version information
````
rustc 1.83.0-dev
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.83.0-dev
LLVM version: 19.1.0

````

### LINKED ISSUE #130858: ICE when specifying wrong return type for closure in static item
This is a pretty minor ICE that I can only reproduce in code that shouldn't compile anyway. If I specify the wrong return type for a closure used in a static item, I get an ICE instead of a normal build error.

Possibly related to https://github.com/rust-lang/rust/issues/125655

### Code
Add this to any `main.rs`:

```Rust
use std::sync::LazyLock;
static FOO: LazyLock<i64> = LazyLock::new(|| -> String { 3 });
```

### Meta

`rustc --version --verbose`:
```
rustc 1.81.0 (eeb90cda1 2024-09-04)
binary: rustc
commit-hash: eeb90cda1969383f56a2637cbd3037bdf598841c
commit-date: 2024-09-04
host: aarch64-apple-darwin
release: 1.81.0
LLVM version: 18.1.7
```

I can also reproduce this on latest nightly (`rustc 1.83.0-nightly (363ae4188 2024-09-24)` I believe, ran it with `cargo +nightly run`).

### Error output

```
error: internal compiler error: compiler/rustc_middle/src/ty/context.rs:3016:21: No bound vars found for HirId(DefId(0:4 ~ ice_repro[b95c]::FOO).0) (static FOO)

thread 'rustc' panicked at compiler/rustc_middle/src/ty/context.rs:3016:21:
Box<dyn Any>
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: <rustc_errors::diagnostic::BugAbort as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
   2: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>::{closure#0}
   3: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}
   4: rustc_middle::ty::context::tls::with_context_opt::<rustc_middle::ty::context::tls::with_opt<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
   5: rustc_middle::util::bug::bug_fmt
   6: <rustc_middle::ty::context::TyCtxt>::late_bound_vars
   7: <rustc_hir_typeck::fn_ctxt::FnCtxt>::suggest_missing_return_type
   8: <rustc_hir_typeck::coercion::CoerceMany<rustc_hir::hir::Arm>>::report_return_mismatched_types
   9: <rustc_hir_typeck::coercion::CoerceMany<&rustc_hir::hir::Expr>>::coerce_inner::<<rustc_hir_typeck::fn_ctxt::FnCtxt>::check_block_with_expected::{closure#0}::{closure#1}>
  10: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_block_with_expected
  11: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  12: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_return_expr
  13: rustc_hir_typeck::check::check_fn
  14: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  15: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  16: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_argument_types
  17: <rustc_hir_typeck::fn_ctxt::FnCtxt>::confirm_builtin_call
  18: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  19: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  20: rustc_hir_typeck::typeck
      [... omitted 2 frames ...]
  21: rustc_hir_analysis::check::check::check_item_type
  22: rustc_hir_analysis::check::wfcheck::check_well_formed
      [... omitted 2 frames ...]
  23: <rustc_data_structures::sync::parallel::ParallelGuard>::run::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_data_structures::sync::parallel::disabled::try_par_for_each_in<&[rustc_hir::hir::ImplItemId], rustc_span::ErrorGuaranteed, <rustc_middle::hir::ModuleItems>::par_impl_items<rustc_hir_analysis::check::wfcheck::check_mod_type_wf::{closure#1}>::{closure#0}>::{closure#0}::{closure#0}::{closure#0}>
  24: rustc_hir_analysis::check::wfcheck::check_mod_type_wf
      [... omitted 2 frames ...]
  25: <rustc_data_structures::sync::parallel::ParallelGuard>::run::<(), rustc_data_structures::sync::parallel::disabled::par_for_each_in<&[rustc_hir::hir_id::OwnerId], <rustc_middle::hir::map::Map>::par_for_each_module<rustc_hir_analysis::check_crate::{closure#0}::{closure#0}>::{closure#0}>::{closure#0}::{closure#0}::{closure#0}>
  26: <rustc_session::session::Session>::time::<(), rustc_hir_analysis::check_crate::{closure#0}>
  27: rustc_hir_analysis::check_crate
  28: rustc_interface::passes::analysis
      [... omitted 2 frames ...]
  29: <rustc_interface::queries::QueryResult<&rustc_middle::ty::context::GlobalCtxt>>::enter::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}::{closure#1}::{closure#5}>
  30: <rustc_interface::interface::Compiler>::enter::<rustc_driver_impl::run_compiler::{closure#0}::{closure#1}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_span::ErrorGuaranteed>>
  31: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>
  32: rustc_span::create_session_globals_then::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_interface::util::run_in_thread_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.81.0 (eeb90cda1 2024-09-04) running on aarch64-apple-darwin

note: compiler flags: --crate-type bin -C embed-bitcode=no -C debuginfo=2 -C split-debuginfo=unpacked -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [typeck] type-checking `FOO`
#1 [check_well_formed] checking that `FOO` is well-formed
#2 [check_mod_type_wf] checking that types are well-formed in top-level module
#3 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `ice-repro` (bin "ice-repro")
```

</p>
</details>

### PR DESCRIPTION
**Pull request by fmease**
Fixes #130858.
Fixes #125655.
Fixes #130391.
Fixes #130663.

r? compiler-errors

### CODE REVIEW

File: tests/ui/closures/closure-return-type-mismatch.rs [hunk 1]
```diff
@@ -15,3 +15,8 @@ fn main() {
         b
     };
 }
+
```
Comment (fmease) [inline tests/ui/closures/closure-return-type-mismatch.rs:18]: Couldn't find a better place to put id, I didn't want to create a new file for this.

**Review by compiler-errors — APPROVED** [review]

**Comment by fmease** [comment]
Right, this also fixes #130391. I'll add another test.
@bors r-

**Comment by fmease** [comment]
Ah, and #130663 ^^. Same reason.
@bors r-

**Pull request merged**